src/editor/controls/GitHubRepositoryControl.tsx

```diff
@@ -0,0 +1,119 @@
+import {BaseControl, Button, TextControl,} from '@wordpress/components';
+import {useEffect, useRef} from '@wordpress/element';
+import {__} from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
+
+interface GitHubRepositoryControlProps {
+    value: string;
+    onChange: (value: string) => void;
+}
+
+export const GitHubRepositoryControl = ({value, onChange, onCodeFetched}: GitHubRepositoryControlProps) => {
+    const inputRef = useRef<HTMLInputElement>(null);
+
+    useEffect(() => {
+        if (isValidUrl(value)) {
+            fetchFile(value).then((code) => {
+                if (value.indexOf("#L")) {
+                    onCodeFetched({ code, lineNumbers: extractLineNumbers(value) });
+                } else {
+                    onCodeFetched({ code });
+                }
+            });
+
+            inputRef.current
+                ?.querySelector('input')
+                ?.classList.remove('cbp-input-error');
+        } else {
+            inputRef.current
+                ?.querySelector('input')
+                ?.classList.add('cbp-input-error');
+        }
+    }, [value]);
```

Comment:
This is good but instead of this we should use `swr` to fetch the code. That way it comes with auto sync, re-try, and error handling baked in. It won't be a big change but i can write up an example if you'd like